core/application/impl/app_state_manager_impl.cpp

```diff
@@ -0,0 +1,169 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "application/impl/app_state_manager_impl.hpp"
+
+#include <csignal>
+#include <functional>
+
+namespace kagome {
+
+  std::weak_ptr<AppStateManager> AppStateManagerImpl::wp_to_myself;
+
+  AppStateManagerImpl::AppStateManagerImpl()
+      : logger_(common::createLogger("Application")) {
+    struct sigaction act {};
+    memset(&act, 0, sizeof(act));
+    act.sa_handler = shuttingDownSignalsHandler;  // NOLINT
+    sigset_t set;
+    sigemptyset(&set);
+    sigaddset(&set, SIGINT);
+    sigaddset(&set, SIGTERM);
+    sigaddset(&set, SIGQUIT);
+    act.sa_mask = set;
+    sigaction(SIGINT, &act, nullptr);
+    sigaction(SIGTERM, &act, nullptr);
+    sigaction(SIGQUIT, &act, nullptr);
+    sigprocmask(SIG_UNBLOCK, &act.sa_mask, nullptr);
+  }
+
+  AppStateManagerImpl::~AppStateManagerImpl() {
+    struct sigaction act {};
+    memset(&act, 0, sizeof(act));
+    act.sa_handler = SIG_DFL;  // NOLINT
+    sigset_t set;
+    sigemptyset(&set);
+    sigaddset(&set, SIGINT);
```

Comment:
why are you adding them in destructor? pls explain.

Reply:
After destruct manager, needs to return back previous signal handler. Without it possible undefined behavior. In this case nothing bad right now, but it woild be changed in future.